File: JACO/Ant.java
/*
Name: Jarred Durant
Course: CIS 421 -- Artificial Intelligence
Instructor: Dr. Laura Grabowski
Assignment #4 -- Pathfinding in Middle-Earth with Ant Colony Optimization
*/

import java.util.*;

public class Ant {
  public static double ALPHA;   // weight of pheromone
  public static double BETA;    // weight of heuristic
  public static double RHO;     // trail persistence: 1 - RHO = evap
  public static int Q;          // quantity of pheromone to distribute
  // Ordered list of locations visited
  public static ArrayList<Location> tourMemory;
  // Map of how much pheromone to lay across each road.
  public static Map<Road, Double> pheromoneToLay = new TreeMap<Road, Double>();
  public static Location currentLoc;  // current location


  // Precondition: currentLoc is not null.
  // Postcondition: Move locations; currentLoc is added to tourMemory.
  public static void moveAround(boolean print) {
    while (!currentLoc.name.equals(ACOPathfinder.goalLoc)) {

      tourMemory.add(currentLoc);
      currentLoc = pickNextLocation();
      layPheromone();
      //currentLoc = next;
      if (print) ACOPathfinder.out.append(currentLoc.name + "\n");
      currentLoc.visited = true;
    }
  }

  // If we're "trapped" by being surrounded by visited locations, pretend we
  // never visited any surrounding locations.
  // Precondition: currentLoc is not null
  public static void recoverTrapped() {
      for (Road r : currentLoc.connections)
        r.otherLoc(currentLoc).visited = false;
  }

  // Probabilistically chooses the next location to move to, using the values of
  // ALPHA, BETA, and RHO in accordance with the Ant Colony Optimization
  // algorithm described in the text.
  // Precondition: currentLoc is not null
  public static Location pickNextLocation() {
    TreeMap<Location, Double> probabilities = new TreeMap<Location, Double>();
    double sigma = 0.0;
    for (Road r : currentLoc.connections) {
      if (r.otherLoc(currentLoc).name.equals(ACOPathfinder.goalLoc))
        return r.otherLoc(currentLoc);
      while (checkTrapped()) recoverTrapped();
      if (!r.otherLoc(currentLoc).visited) {
        double t = r.pheromone;
        t = Math.pow(t, ALPHA);
        double h = 1.0 / r.distance;
        h = Math.pow(h, BETA);
        sigma += t * h;
      }
    }

    for (Road r : currentLoc.connections) {
      if (!r.otherLoc(currentLoc).visited) {
        double t = r.pheromone;
        t = Math.pow(t, ALPHA);
        double h = 1.0 / r.distance;
        h = Math.pow(h, BETA);
        double prob = t * h / sigma;
        probabilities.put(r.otherLoc(currentLoc), prob);
      }
    }

    Random randGen = new Random();
    double rand = randGen.nextDouble();
    double runningTotal = 0;
    for (Location l : probabilities.keySet()) {
      runningTotal += probabilities.get(l);
      if (runningTotal > rand) return l;
    }

    Iterator<Location> i = probabilities.keySet().iterator();

    return i.next(); // should never happen
  }

  // Check to see if we're "trapped" (surrounded by visited locations)
  public static boolean checkTrapped() {
     for (Road r : currentLoc.connections) {
       if (!r.otherLoc(currentLoc).visited) return false;
     }
     return true;
  }

  // Lay pheromone across roads
  public static void layPheromone() {
    for (int i = 0; i < tourMemory.size() - 2; i++) {
      Location l1 = tourMemory.get(i);
      Location l2 = tourMemory.get(i + 1);
      Road r = null;
      for (Road thisRoad : ACOPathfinder.roads) {
        if ((thisRoad.loc1.equals(l1) && thisRoad.loc2.equals(l2))
                    || (thisRoad.loc1.equals(l2) && thisRoad.loc2.equals(l1))) {
          r = thisRoad;
          break;
        }
      }
      pheromoneToLay.put(r, (double) Q / tourMemory.size() - 1);
    }
  }

  // Constructor: self-explanatory, I hope
  public Ant(Location l, double inAlpha, double inBeta, double inRho, int inQ) {
    tourMemory = new ArrayList<Location>();
    currentLoc = l;
    ALPHA = inAlpha;
    BETA = inBeta;
    RHO = inRho;
    Q = inQ;
  }
}

File: src/ant.rb
class Ant
  attr_accessor :tour_memory, :start, :goal, :current_location, :pheromone_flavor

  def initialize(start, goal, pheromone_flavor, map)
    @tour_memory = []
    @goal = goal
    @current_location = start
    @start = start
    @next_location = nil
    @pheromone_flavor = pheromone_flavor
    @map = map
  end

  def to_json(*args)
    @pheromone_flavor.to_s
  end

  def travel
    while @current_location != @goal do
      @tour_memory << @current_location
      @current_location = pick_next_location
     end
  end
  
  def lay_pheromone
    if @tour_memory.size == 2
      loc = @tour_memory[0]
      loc2 = @tour_memory[1]
      road = @map.find_road loc, loc2
      road.pheromones[@pheromone_flavor] += Q
    else
      @tour_memory.each_with_index do |loc, ix|
        if ix < @tour_memory.size - 1
          loc2 = @tour_memory[ix + 1]
          road = @map.find_road loc, loc2
          road.pheromones[@pheromone_flavor] += Q / (@tour_memory.size - 1)
        end
      end
    end
  end

  def is_trapped?
    @current_location.connections.each do |road|
      return false unless road.other_loc(@current_location).visited
    end
  true
  end

  def recover_from_trap
    output.write "I'm in #{@current_location} - recovering from trap!"
    @current_location.connections.each do |road|
      road.other_loc(current_location).visited = false
    end
  end

  def pick_next_location
    probabilities = {}
    sigma = 0.0
    @current_location.connections.each do |road|
      return goal if road.other_loc(current_location) == goal
      recover_trapped while is_trapped?
      unless road.other_loc(current_location).visited
        t = road.pheromones[pheromone_flavor] ** @map.alpha
        h = (1.0 / road.distance) ** @map.beta
        sigma += t * h
      end
    end

    @current_location.connections.each do |road|
      unless road.other_loc(current_location).visited
        t = road.pheromones[pheromone_flavor] ** @map.alpha
        h = (1.0 / road.distance) ** @map.beta
        prob = (t * h) / (sigma + 0.0000001)
        probabilities[road.other_loc(current_location)] = prob
      end
    end

    rand = Random.rand
    running_total = 0
    probabilities.each do |loc, prob|
      running_total += prob
      return loc if running_total > rand
    end
  end
end